### PR TITLE
[feature/gke-private-saas-cluster-v1] Add the first cut of GKE cluster creation

### PR DESCRIPTION
**Pull request by aravind-opsverse**
**What is done?**
1. Basic cluster creation
2. Access and network connectivity working as expected

**What is pending?**
1. GCS bucket and IAM role/Service Account creation
3. Remote state management
4. Hardening of security and network rules
5. IAM and User management
6. Simplify network and subnetwork sections

**How was this code tested?**
Planned and applied the code to create the cluster. Ran self-hosted `ObserveNow` instance to ensure networking works as expected.

```
➜  private-saas git:(main) ✗ terraform apply "gke-plan-new.tfplan"                                                           
module.opsverse-gke-cluster.random_string.cluster_service_account_suffix[0]: Creating...
module.opsverse-gke-cluster.random_string.cluster_service_account_suffix[0]: Creation complete after 0s [id=crok]
module.opsverse-gke-cluster.google_service_account.cluster_service_account[0]: Creating...
module.opsverse-gke-cluster-network.module.vpc.google_compute_network.network: Creating...
module.opsverse-gke-cluster.random_shuffle.available_zones[0]: Creating...
module.opsverse-gke-cluster.random_shuffle.available_zones[0]: Creation complete after 0s [id=-]
module.opsverse-gke-cluster.google_service_account.cluster_service_account[0]: Still creating... [10s elapsed]
module.opsverse-gke-cluster-network.module.vpc.google_compute_network.network: Still creating... [10s elapsed]
module.opsverse-gke-cluster.google_service_account.cluster_service_account[0]: Creation complete after 14s [id=projects/opsverse-dev-poc/serviceAccounts/tf-gke-opsverse-gke-cl-crok@opsverse-dev-poc.iam.gserviceaccount.com]
module.opsverse-gke-cluster.google_project_iam_member.cluster_service_account-resourceMetadata-writer[0]: Creating...
module.opsverse-gke-cluster.google_project_iam_member.cluster_service_account-nodeService_account[0]: Creating...
module.opsverse-gke-cluster.google_project_iam_member.cluster_service_account-metric_writer[0]: Creating...
module.opsverse-gke-cluster-network.module.vpc.google_compute_network.network: Still creating... [20s elapsed]
module.opsverse-gke-cluster.google_project_iam_member.cluster_service_account-nodeService_account[0]: Still creating... [10s elapsed]
module.opsverse-gke-cluster.google_project_iam_member.cluster_service_account-resourceMetadata-writer[0]: Still creating... [10s elapsed]
module.opsverse-gke-cluster.google_project_iam_member.cluster_service_account-metric_writer[0]: Still creating... [10s elapsed]
module.opsverse-gke-cluster.google_project_iam_member.cluster_service_account-resourceMetadata-writer[0]: Creation complete after 13s [id=opsverse-dev-poc/roles/stackdriver.resourceMetadata.writer/serviceAccount:tf-gke-opsverse-gke-cl-crok@opsverse-dev-poc.iam.gserviceaccount.com]
module.opsverse-gke-cluster.google_project_iam_member.cluster_service_account-nodeService_account[0]: Creation complete after 13s [id=opsverse-dev-poc/roles/container.defaultNodeServiceAccount/serviceAccount:tf-gke-opsverse-gke-cl-crok@opsverse-dev-poc.iam.gserviceaccount.com]
module.opsverse-gke-cluster.google_project_iam_member.cluster_service_account-metric_writer[0]: Creation complete after 14s [id=opsverse-dev-poc/roles/monitoring.metricWriter/serviceAccount:tf-gke-opsverse-gke-cl-crok@opsverse-dev-poc.iam.gserviceaccount.com]
module.opsverse-gke-cluster-network.module.vpc.google_compute_network.network: Still creating... [30s elapsed]
module.opsverse-gke-cluster-network.module.vpc.google_compute_network.network: Creation complete after 34s [id=projects/opsverse-dev-poc/global/networks/opsverse-gke-cluster-apps-new]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-private-b"]: Creating...
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/popsverse-gke-cluster-apps-public-c"]: Creating...
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-private-c"]: Creating...
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-public-a"]: Creating...
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-public-b"]: Creating...
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-private-a"]: Creating...
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/popsverse-gke-cluster-apps-public-c"]: Still creating... [10s elapsed]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-private-b"]: Still creating... [10s elapsed]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-private-c"]: Still creating... [10s elapsed]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-public-b"]: Still creating... [10s elapsed]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-public-a"]: Still creating... [10s elapsed]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-private-a"]: Still creating... [10s elapsed]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/popsverse-gke-cluster-apps-public-c"]: Creation complete after 17s [id=projects/opsverse-dev-poc/regions/us-east4/subnetworks/popsverse-gke-cluster-apps-public-c]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-private-b"]: Still creating... [20s elapsed]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-private-c"]: Still creating... [20s elapsed]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-public-a"]: Still creating... [20s elapsed]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-public-b"]: Still creating... [20s elapsed]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-private-a"]: Still creating... [20s elapsed]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-public-a"]: Creation complete after 28s [id=projects/opsverse-dev-poc/regions/us-east4/subnetworks/opsverse-gke-cluster-apps-public-a]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-private-a"]: Creation complete after 28s [id=projects/opsverse-dev-poc/regions/us-east4/subnetworks/opsverse-gke-cluster-apps-private-a]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-public-b"]: Creation complete after 29s [id=projects/opsverse-dev-poc/regions/us-east4/subnetworks/opsverse-gke-cluster-apps-public-b]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-private-c"]: Creation complete after 29s [id=projects/opsverse-dev-poc/regions/us-east4/subnetworks/opsverse-gke-cluster-apps-private-c]
module.opsverse-gke-cluster-network.module.subnets.google_compute_subnetwork.subnetwork["us-east4/opsverse-gke-cluster-apps-private-b"]: Creation complete after 29s [id=projects/opsverse-dev-poc/regions/us-east4/subnetworks/opsverse-gke-cluster-apps-private-b]
module.opsverse-gke-cluster.google_container_cluster.primary: Creating...
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [10s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [20s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [30s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [40s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [50s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [1m0s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [1m10s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [1m20s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [1m30s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [1m40s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [1m50s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [2m0s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [2m10s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [2m20s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [2m30s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [2m40s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [2m50s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [3m0s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [3m10s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [3m20s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [3m30s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [3m40s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [3m50s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [4m0s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [4m10s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Still creating... [4m20s elapsed]
module.opsverse-gke-cluster.google_container_cluster.primary: Creation complete after 4m22s [id=projects/opsverse-dev-poc/locations/us-east4/clusters/opsverse-gke-cluster]
module.opsverse-gke-cluster.google_container_node_pool.pools["node-pool"]: Creating...
module.opsverse-gke-cluster.google_container_node_pool.pools["node-pool"]: Still creating... [10s elapsed]
module.opsverse-gke-cluster.google_container_node_pool.pools["node-pool"]: Still creating... [20s elapsed]
module.opsverse-gke-cluster.google_container_node_pool.pools["node-pool"]: Still creating... [30s elapsed]
module.opsverse-gke-cluster.google_container_node_pool.pools["node-pool"]: Still creating... [40s elapsed]
module.opsverse-gke-cluster.google_container_node_pool.pools["node-pool"]: Still creating... [50s elapsed]
module.opsverse-gke-cluster.google_container_node_pool.pools["node-pool"]: Still creating... [1m0s elapsed]
module.opsverse-gke-cluster.google_container_node_pool.pools["node-pool"]: Still creating... [1m10s elapsed]
module.opsverse-gke-cluster.google_container_node_pool.pools["node-pool"]: Still creating... [1m20s elapsed]
module.opsverse-gke-cluster.google_container_node_pool.pools["node-pool"]: Creation complete after 1m22s [id=projects/opsverse-dev-poc/locations/us-east4/clusters/opsverse-gke-cluster/nodePools/node-pool]
module.opsverse-gke-cluster-auth.data.google_client_config.provider: Reading...
module.opsverse-gke-cluster-auth.data.google_container_cluster.gke_cluster: Reading...
module.opsverse-gke-cluster-auth.data.google_client_config.provider: Read complete after 0s [id=projects/<null>/regions/<null>/zones/<null>]
module.opsverse-gke-cluster-auth.data.google_container_cluster.gke_cluster: Read complete after 9s [id=projects/opsverse-dev-poc/locations/us-east4/clusters/opsverse-gke-cluster]
local_file.kubeconfig-opsverse-gke-cluster: Creating...
local_file.kubeconfig-opsverse-gke-cluster: Creation complete after 0s [id=13d90bc1b1e4a44847f83b1ce08f949b8be09ceb]

Apply complete! Resources: 16 added, 0 changed, 0 destroyed.

Outputs:

network_name = "opsverse-gke-cluster-apps-new"
subnets_names = [
  "opsverse-gke-cluster-apps-private-a",
  "opsverse-gke-cluster-apps-private-b",
  "opsverse-gke-cluster-apps-private-c",
  "opsverse-gke-cluster-apps-public-a",
  "opsverse-gke-cluster-apps-public-b",
  "opsverse-gke-cluster-apps-public-c",
]
```

**Stack launch and testing**
Screenshot 2024-08-30 at 10 34 58 PM

```
➜  ~ dig <hostname>.opsverse.cloud

; <<>> DiG 9.10.6 <<>> <hostname>.opsverse.cloud
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 36008
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;<hostname>.opsverse.cloud.	IN	A

;; ANSWER SECTION:
<hostname>.opsverse.cloud. 239 IN	A	34.48.xx.x

;; Query time: 18 msec
;; SERVER: 218.248.112.1#53(218.248.112.1)
;; WHEN: Fri Aug 30 22:35:44 IST 2024
;; MSG SIZE  rcvd: 72
```

### CODE REVIEW

File: gcp/opsverse-gke/main.tf [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+      // nodes per location (this multiplied by number
+      // of locations will be the current node count)
+      node_count           = 1
+      initial_node_count   = 3
+      min_count            = 3
+      max_count            = 3
+      disk_size_gb         = 30
+      autoscaling          = false 
+      auto_upgrade         = true
```
Comment (navinpai) [inline gcp/opsverse-gke/main.tf]: Set this to false.

File: gcp/opsverse-gke/main.tf [hunk 2]
```diff
@@ -0,0 +1,45 @@
+resource "local_file" "kubeconfig-opsverse-gke-cluster" {
```
Comment (navinpai) [inline gcp/opsverse-gke/main.tf:1]: Please add a `README`

File: gcp/opsverse-gke/network.tf [hunk 3]
```diff
@@ -0,0 +1,100 @@
+module "opsverse-gke-cluster-network" {
+  source       = "terraform-google-modules/network/google"
+  version      = "9.1.0"
+  project_id   = var.gcp_project_id
+  network_name = "opsverse-gke-cluster-apps-new"
+  subnets = [
```
Comment (navinpai) [inline gcp/opsverse-gke/network.tf]: Make these configurable like https://github.com/OpsVerseIO/private-saas/blob/main/aws/opsverse-eks-iam/vars.tfvars#L18

File: gcp/opsverse-gke/main.tf [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  name                     = var.cluster_name
+  regional                 = true
+  region                   = var.gcp_region
+  network                  = module.opsverse-gke-cluster-network.network_name
+  subnetwork               = module.opsverse-gke-cluster-network.subnets_names[0]
+  ip_range_pods            = "ip-range-pods" 
+  ip_range_services        = "ip-range-services" 
+  node_pools = [
+    {
+      name                 = "node-pool"
```
Comment (navinpai) [inline gcp/opsverse-gke/main.tf]: Add `opsverse` to the name

File: gcp/opsverse-gke/main.tf (hunk 1 above)
Comment (aravind-opsverse) [reply gcp/opsverse-gke/main.tf]: Done

File: gcp/opsverse-gke/main.tf (hunk 4 above)
Comment (aravind-opsverse) [reply gcp/opsverse-gke/main.tf]: Done

File: gcp/opsverse-gke/main.tf (hunk 2 above)
Comment (aravind-opsverse) [reply gcp/opsverse-gke/main.tf:1]: Done

File: gcp/opsverse-gke/network.tf (hunk 3 above)
Comment (aravind-opsverse) [reply gcp/opsverse-gke/network.tf]: Done

**Review by navinpai — APPROVED** [review]

**Pull request merged**